### PR TITLE
Collapsing nodes

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
Code has been mostly cleaned up. Some places that particularly require more work are marked with todos.
* More extensive tests are needed.

### Checklist
Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [ ] All code conforms to the [Rust](https://github.com/luna/enso/blob/main/docs/style-guide/rust.md).
- [ ] All code has been tested where possible.
- [ ] All code has been profiled where possible.

### CODE REVIEW

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 1]
```diff
@@ -1187,7 +1187,12 @@ impl Block<Ast> {
         let empty_lines = Vec::new();
         let first_line  = BlockLine::new(first_line.clone_ref());
         let lines       = tail_lines.iter().cloned().map(BlockLine::new).collect();
-        let is_orphan   = tail_lines.is_empty();
+        // We virtually never want block to be an orphan (i.e. start with a first line attached
+        // to whatever AST was there before). While it may seem tempting to have this for single
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/lib.rs:1191]: I don't understand this "start with a first line attached to whatever AST was there before" I see the comment that orphan is "Block start with a leading newline" what does not make sense to me as well, especially comparing with the function docs.

**Review by farmaazon — COMMENTED** [review]
As far I think this is neatly done. No big issues from my side, I wait for cleaned-up code.

File: src/rust/ide/lib/ast/impl/src/lib.rs (hunk 1 above)
Comment (mwu-tow) [reply src/rust/ide/lib/ast/impl/src/lib.rs:1191]: The documentation for the block is incorrect. The leading newline is present when the `is_orphan` is false, not the other way around.
As for clarifying this comment, let's discuss this tomorrow.

**Comment by mwu-tow** [comment]
@farmaazon 
I've updated the PR with heavy refactoring of the `collapse` module, better test coverage and all missing logic.
Still there are a few docstrings to be improved, more tests to be added and a few TODO notes left.
WIth that caveats, the PR should be ready for the review.

**Pull request merged**